module/zfs/dnode.c

```diff
@@ -1081,31 +1081,40 @@ dnode_set_slots(dnode_children_t *children, int idx, int slots, void *ptr)
 	}
 }
 
+static boolean_t
+dnode_check_evictable(dnode_t *dn)
+{
+	boolean_t can_evict;
+
+	mutex_enter(&dn->dn_mtx);
+	can_evict = (dn->dn_type == DMU_OT_NONE && !DNODE_IS_DIRTY(dn));
+	mutex_exit(&dn->dn_mtx);
+	return (can_evict);
+}
```

Comment:
I'm not convinced it's worth adding a function for this.  How about, this change in `dnode_check_slots_free`?

```diff
-			dmu_object_type_t type = dn->dn_type;
+			boolean_t free = (dn->dn_type == DMU_OT_NONE && !DNODE_IS_DIRTY(dn));
```